### PR TITLE
Feature: UIControllers

### PR DESCRIPTION
**Pull request by shakyShane**
**Reviewer:** @GioSensation @jonathanKingston 
**Asana:** https://app.asana.com/0/0/1202241610344476/f

## Description

To expand to more platforms, our current architecture needs a bit of a reshuffle to incorporate some "Separation of concerns"

This PR introduces `UIControllers` which separates the tooltip handling logic from the devices. A follow-up PR will heavily use this to bring Logins+ to 3 new platforms. 

This change is *mostly* just de-coupling tooltip logic from devices, which naturally makes the diff look larger - in reality very little code has actually changed :)

## Steps to test

The following **manual** tests will be conducted by me - the PR can still be reviewed whilst this is ongoing though 🚀

- [x] macOS (with overlay)
- [x] macOS (without overlay)
- [x] Android (email protection only)
- [x] iOS (email protection only)
- [x] Extension (email protection only)

**before**

Untitled Diagram drawio (3)

**after**

Untitled Diagram drawio (4)

**future**

Untitled Diagram drawio (5)

### CODE REVIEW

**Review by GioSensation — APPROVED** [review]
Just a few minor comments and questions. Otherwise it seems fine. I still can't help but wonder whether we could have gone for a simpler solution. 🤔

File: src/UI/controllers/OverlayUIController.js [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    /**
+     * @param {import('./UIController').AttachArgs} args
+     */
+    attach (args) {
+        const {getPosition, topContextData, click, input} = args
+        let delay = 0
+        if (!click && !this.elementIsInViewport(getPosition())) {
+            input.scrollIntoView(true)
```
[the remark this answers is not in the document]
Comment (shakyShane) [reply src/UI/controllers/OverlayUIController.js:54]: Yep, it was there before: https://github.com/duckduckgo/duckduckgo-autofill/pull/155/files#diff-39cb8c42a25de5dce09b5fc0aacf6aca587b22abc2de6a9c985a36f9f6b9242dL169-L170

File: src/UI/controllers/NativeUIController.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * ```
+ */
+export class NativeUIController extends UIController {
+    /**
+     * @param {import('./UIController').AttachArgs} _args
+     */
+    attach (_args) {
+        // const {form, input, device} = args
+        // const inputType = getInputType(input)
+        // const mainType = getMainTypeFromType(inputType)
```
[the remark this answers is not in the document]
Comment (shakyShane) [reply src/UI/controllers/NativeUIController.js:23]: removed

File: src/DeviceInterface/AppleOverlayDeviceInterface.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
+import {AppleDeviceInterface} from './AppleDeviceInterface'
+import {CSS_STYLES} from '../UI/styles/styles'
+import {HTMLTooltipUIController} from '../UI/controllers/HTMLTooltipUIController'
+
+/**
+ * This is here to encapsulate
```
[the remark this answers is not in the document]
Comment (shakyShane) [reply src/DeviceInterface/AppleOverlayDeviceInterface.js:6]: Added a bunch more comments :)

File: src/DeviceInterface/AppleDeviceInterface.js [hunk 4]
```diff
@@ -345,6 +237,64 @@ class AppleDeviceInterface extends InterfacePrototype {
...
+                    break
+                }
+            }
+            poll()
+        })
+    }
+    /**
+     * on macOS we try to detect if a click occurred withing a form
+     */
+    onPointerDown (e) {
```
[the remark this answers is not in the document]
Comment (shakyShane) [reply src/DeviceInterface/AppleDeviceInterface.js:282]: Yep, agreed :)

File: src/UI/controllers/OverlayUIController.js [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+            }
+        }
+        this._options.remove()
+            .catch(e => console.error('Could not close parent', e))
+        this.#state = 'idle'
+        this._removeListeners()
+    }
+}
+
+/**
```
[the remark this answers is not in the document]
Comment (shakyShane) [reply src/UI/controllers/OverlayUIController.js:173]: Moved back to the top

**Comment by shakyShane** [comment]
I have addressed the feedback above, and performed the manual tests across all of our supported platforms, time to merge :)

**Pull request merged**